plc4j/api/src/main/java/org/apache/plc4x/java/api/EventPlcConnection.java

```diff
@@ -0,0 +1,34 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.plc4x.java.api;
+
+import org.apache.plc4x.java.api.listener.EventListener;
+
+/**
+ * An extended kind of connection which can broadcast events.
+ *
+ * Particular type of event might vary, it might be a connection state or message exchange.
+ */
+public interface EventPlcConnection extends PlcConnection {
```

Comment:
I think as this is in the API module, there shouldn't be an event-emitting set of connections and non emitting ones. I would rather extend the PlcConnection with these two methods and have them implemented in all drivers.

Reply:
And in addition, we could use the opportunity to add more events ... like "Connected" event "ConnectionFailed", "Identified" (fired as soon as a connection has identified the remote PLC type (Might be interesting for the users), "MetadataChanged", ...

Reply:
> I think as this is in the API module, there shouldn't be an event-emitting set of connections and non emitting ones

I wasn't sure of that because of amount of changes it will force in connection pool module. Logic in pooled connections is a bit different and I am not entirely sure if I will be able to deliver that.

Reply:
Well if you change the core of things it is natural that they need changes in a lot of places ;-) ... the later you introduce the change, the more you need to change. But I would not be in favor of a change that makes our uses have to do instanceof stuff in order to use the API. 

You wouldn't have to do all on your own. We could probably all do our part. I do think that some sort of "ups I think I just lost the connection" mechanism is indeed very important.

Reply:
I mean @splatch has a valid point regarding the pool, so maybe we can merge these to together later when we have full support. Otherwise this feature would be blocked till we have support in pool for example. From an interface perspective this should be fine as we don't expose the api yet when it's not guaranteed to work the intended way. Once this is validated we can just pull the methods up into the PlcConnection. Implementation wise merging later should have almost now effect when done later then.